experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+
+
+def materialize_module(
+    mod: nn.Module,
+    tensor_materializer: Callable[[FakeTensor], torch.Tensor] = _materialize_tensor,
+) -> Tuple[nn.Module, dict[torch.Tensor, torch.Tensor]]:
+    def to_real_tensor(m: nn.Module, key: str, t: T) -> Union[T, torch.Tensor, nn.Parameter]:
+        if isinstance(t, nn.Parameter):
+            return nn.Parameter(tensor_materializer(t.data, t))
+        else:
+            assert isinstance(t, torch.Tensor)
+            return tensor_materializer(t)
+
+    def materialize_this_module(m: nn.Module) -> None:
+        for key, param in m._parameters.items():
+            assert isinstance(param, nn.Parameter)
+            if True:  # if isinstance(param.data, FakeTensor):
+                dparam = to_real_tensor(m, key, param)
+                assert isinstance(dparam, nn.Parameter)
+                m.register_parameter(key, dparam)
+        for key, buffer in m._buffers.items():
+            if True:  # if isinstance(buffer, FakeTensor):
+                m._buffers[key] = to_real_tensor(m, key, buffer)
+
+    materialize_this_module(mod)
+    for name, submod in mod.named_modules():
+        materialize_this_module(submod)
+
+    return mod
+
+
+def get_example_shard(tensor, mesh, default=None, fn=torch.randn):
+    if not isinstance(tensor, Tensor):
+        return tensor
+    elif isinstance(tensor, DTensor):
+        return tensor.to_local()
+    elif hasattr(tensor, '_placements'):
+        inp_spec = DTensorSpec(mesh, tensor._placements, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    elif default is not None:
+        inp_spec = DTensorSpec(mesh, default, tensor.shape, tensor.ndim)
+        return fn(inp_spec.local_shape, device=tensor.device)
+    raise NotImplementedError(f'Dont know how to get a shard of tensor {tensor}')
+
+
+
+def make_inductor_fn(fn, example_inputs, default_shard_schema=None):
+    """
+    Use inductor to compile the function fn. 
+    Uses make_functional_fx to trace it passing example_inputs.
+    """
+    fnfx, attr_vals, unused_input_idx = make_functional_fx(fn)(*example_inputs)
+
+    used_args = prepare_args(attr_vals, unused_input_idx, example_inputs)
+
+    # materialize models + SPMD expansion. must be outside of fake mode
+    with maybe_disable_fake_tensor_mode():
+        if default_shard_schema is None:
+            tensor_materializer = _materialize_tensor
+        else:
+            tensor_materializer = partial(
+                _materialize_dtensor,
+                default_shard_schema.mesh,
+                default_shard_schema.placements,
+            )
+        # In the following lines we perform utterly black magic in order to
+        # materialize our models. It special cases models and optimizers
+        # and assumes that all parameters contained in optimziers are model
+        # parameters.
+        tracking_materializer = TrackingMaterializer(_materialize_tensor)
```

Comment:
should pass in `tensor_materializer` instead?

Reply:
oh wait yeah this is a bug !

Reply:
after fixing this, turns out warmup_optimizer() i not working for adam with DTensor, because it creates a torch.zeros() internally to track training step. Should we have DTensor mode for these things?

Reply:
@wanchaol I implemented a simple/hacky DTensorMode to deal with this.